BLD: streamlined library names in site.cfg sections

zerothi opened the pull request:
This unifies the names for retrieving libraries in sections.
Previously one could only get umfpack libraries using:

    [umfpack]
    umfpack_libs = ...

while now it prefers to use

    [umfpack]
    libraries = ...

(but defaults to use umfpack_libs).
Some libraries is system_info uses libraries whereas others were using
`*_libs` as keys. All future developments I think should retain the
libraries keyword, and only if applicable for sub-packages it should
be deferred to more `_libs` entries.

This commit also fixes CBLAS linkage if the user has supplied a
sufficient library. I.e. it first checks whether the CBLAS interface
is present *as is*, then tries to add cblas, etc.
It should at least now work for more diverse libraries.
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1131,7 +1133,9 @@ def calc_info(self):
                                    self._lib_names + self._lib_atlas)
+        atlas_libs = self.get_libs('libraries', atlas_libs)
         lapack_libs = self.get_libs('lapack_libs', self._lib_lapack)
+        lapack_libs = self.get_libs('lapack_libraries', lapack_libs)
```
Why a third spelling?

zerothi replied:
Agreed!

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1716,34 +1725,43 @@ def has_cblas(self, info):
+            except (distutils.ccompiler.CompileError, distutils.ccompiler.LinkError):
+                res = None
+                # skip to finally clause, we can't compile anyway!
+                raise PassToFinal
```
Just use `return None` here

zerothi replied:
Indeed that is cleaner!

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1724,26 +1732,26 @@ def has_cblas(self, info):
-                    res = "cblas"
+                    res = libs
+                    # This breaks the for loop
+                    break
```
Just use `return libs` here

zerothi replied:
done, this implies that the last `return res` also changes to `return None`, for clarity!

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1683,6 +1689,7 @@ class blas_info(system_info):
     def calc_info(self):
         lib_dirs = self.get_lib_dirs()
         blas_libs = self.get_libs('blas_libs', self._lib_names)
+        blas_libs = self.get_libs('libraries', blas_libs)
```
Does this merge the results or discard one half off them if both are specified?

zerothi replied:
It has `libraries` precedence. So if only one is present, only that one will be used. I.e.:

```bash
[blas]
blas_libs = foo,bar

[blas]
libraries = foo,bar

[blas]
blas_libs = discarded
libraries = foo,bar
```
will all yield the same.

eric-wieser replied:
In that case, you should emit an error if both are specified.

zerothi replied:
If both are specified and we should raise an error it would require some additional restructuring which I don't think is necessary.
I would be against this since many of the `site.cfg` are undocumented features anyways. I.e. people would probably in some overlapping period expect the same `site.cfg` to work for both cases and when the older `*_libs` finally gets deprecated *then* one could do something.

Do you still think both should raise an error?

zerothi replied:
Additionally I think this is of very minor relevance to 99.9% of users since the `site.cfg` is generally generated with one configuration in mind.

eric-wieser replied:
>  are undocumented features anyways

All the more reason to raise a clear error if they do something that's a bad idea

> Do you still think both should raise an error?

Yes, silently discarding information should definitely raise an error.

zerothi replied:
Ok, I'll fix this

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1698,14 +1705,15 @@ def calc_info(self):
         self.set_info(**info)
 
     def has_cblas(self, info):
+        """ changed in 1.17 to return a list instead of a str """
```
Can you move this change to another PR? The changes to this function seem unrelated to the renaming you do above.

zerothi replied:
Ok, I'll do that, I'll need to extract the entire CBLAS change then.

zerothi replied:
See #13161

zerothi committed:
BLD: streamlined library names in site.cfg sections

This unifies the names for retrieving libraries in sections.
Previously one could only get umfpack libraries using:

  [umfpack]
  umfpack_libs = ...

while now it prefers to use

  [umfpack]
  libraries = ...

(but defaults to use umfpack_libs).
Some libraries is system_info uses libraries whereas others were using
*_libs as keys. All future developments I think should retain the
libraries keyword, and only if applicable for sub-packages it should
be deferred to more _libs entries.

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
BLD: streamlined library names in site.cfg sections

This unifies the names for retrieving libraries in sections.
Previously one could only get umfpack libraries using:

  [umfpack]
  umfpack_libs = ...

while now it prefers to use

  [umfpack]
  libraries = ...

(but defaults to use umfpack_libs).
Some libraries is system_info uses libraries whereas others were using
*_libs as keys. All future developments I think should retain the
libraries keyword, and only if applicable for sub-packages it should
be deferred to more _libs entries.

Signed-off-by: Nick Papior <nickpapior@gmail.com>

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1131,6 +1133,7 @@ def calc_info(self):
         atlas_libs = self.get_libs('atlas_libs',
                                    self._lib_names + self._lib_atlas)
+        atlas_libs = self.get_libs('libraries', atlas_libs)
         lapack_libs = self.get_libs('lapack_libs', self._lib_lapack)
```
Why is `lapack_libs` still used as the `x_libs` format, but `atlas_libs` aliased to `libraries`?

zerothi replied:
Yes, this is a bit problematic with ATLAS. ATLAS can provide either BLAS or BLAS+LAPACK(sub) or BLAS+LAPACK, with the middle option being only a subset of routines which are optimized.
I would suggest *any* user to put all libraries (both BLAS and LAPACK) in a single library entry, however it seems that the dev of the `atlas_info` class recommended LAPACK and BLAS libraries to be split. The only case where this would be troublesome is to use ATLAS BLAS but another LAPACK lib.

Basically I think this is why there is an abstraction of another key for only the LAPACK libraries.

So the final atlas section would look like:
```bash
[atlas]
libraries = f77blas,cblas,atlas
lapack_libs = lapack_atlas
```
And then a user could do `get_info('blas_opt').update(get_info('lapack_mkl'))` for instance (dummy example ;) )

rgommers replied:
I think splits are possible, they're different libraries (e.g. BLIS + Flame).

> ATLAS can provide either BLAS or BLAS+LAPACK(sub) or BLAS+LAPACK, with the middle option being only a subset of routines which are optimized.

I have never seen ATLAS used with numpy/scipy as anything other than "all of BLAS + LAPACK".

zerothi replied:
@rgommers agreed, me neither. But the possibility *is* there.

eric-wieser commented on numpy/distutils/system_info.py:
```diff
@@ -1877,10 +1884,8 @@ class blis_info(blas_info):
         lib_dirs = self.get_lib_dirs()
-        blis_libs = self.get_libs('libraries', self._lib_names)
-        if blis_libs == self._lib_names:
-            blis_libs = self.get_libs('blis_libs', self._lib_names)
```
I hadn't realized we already accepted `libraries` elsewhere. I think using this approach with the `if` statement everywhere makes sense to me.

zerothi replied:
Yes, it has been there for some time. But I dislike the above since one could put the same entries in the `libraries` key which would then allow the 2nd key.

In my opinion we should fix a `libraries` preference, and allow the older keywords (as exceptions). Once I have the error handler I will update.

zerothi committed:
ENH: added AliasedOptionError for system_info

This enables one to force users to use only one keyword in
the site.cfg file for distutils.

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
ENH: added AliasedOptionError for system_info

This enables one to force users to use only one keyword in
the site.cfg file for distutils.

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi commented:
@eric-wieser, @rgommers I have marked conversations as resolved. Please reopen if needed!

mattip commented on numpy/distutils/system_info.py:
```diff
@@ -607,6 +611,35 @@ def set_info(self, **info):
+            return options[0]
+
+        # Else we have more than 1 key found
+        raise AliasedOptionError(AliasedOptionError.__doc__.format(section=self.section, options='[{}]'.format(', '.join(options))))
```
80 columns

mattip commented on numpy/distutils/system_info.py:
```diff
@@ -449,6 +449,10 @@ class NotFoundError(DistutilsError):
+class AliasedOptionError(DistutilsError):
+    """Aliases entries in config files should not be existing. In section '{section}' we found multiple of options {options} """
+
+
```
Needs reformat to 80 column width. Also "mulitple of options" might be better as "multiple appearances of option"

mattip commented:
Could use some new tests in `numpy/distutils/tests/test_system_info.py`

zerothi committed:
MAINT: fixed 80 character width and added a test

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
MAINT: fixed 80 character width and added a test

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi commented:
> Could use some new tests in `numpy/distutils/tests/test_system_info.py`

Added!

zerothi committed:
MAINT: fixed PYTHONOPTIMIZE=2 since docstrings are removed

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
MAINT: fixed PYTHONOPTIMIZE=2 since docstrings are removed

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi commented:
@mattip fixed remaining issues!

mattip commented:
It would be nice to document this but ...
* `site.cfg.example` is not linked into the html document tree at all.
* Documentation for `system_info` (which uses and describes `site.cfg`) is in the docstring for `numpy/distutils/system_info.py` but is not linked to the documentation, the closest we get is [`doc/DISTUTILS.rst.txt`](http://www.numpy.org/devdocs/reference/distutils_guide.html#numpy-distutils-system-info-module).

Could you take a look at both those files anyway (`site.cfg.example` and `system_info.py`) and update them for the preferred `libraries=` usage? Extra points if you figure out a nice way to link that into the documentation, but it should probably be a different issue/PR

zerothi committed:
Merge branch 'linalg-libraries' of github.com:zerothi/numpy into linalg-libraries

zerothi committed:
Merge branch 'master' into linalg-libraries

zerothi committed:
MAINT: updated documentation per @mattip request

Signed-off-by: Nick Papior <nickpapior@gmail.com>

mattip commented on site.cfg.example:
```diff
@@ -21,6 +21,8 @@
 #       example, the file "libfoo.so" would become simply "foo".
 #           libraries = lapack,f77blas,cblas,atlas
+#       This setting is available for *all* sections, if not it should be
+#       considered a bug.
```
No need for the ", if not ..."

zerothi replied:
done

mattip commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -7,9 +7,10 @@
 from numpy.distutils import ccompiler, customized_ccompiler
-from numpy.testing import assert_, assert_equal
 from numpy.distutils.system_info import system_info, ConfigParser
+from numpy.distutils.system_info import AliasedOptionError
```
Collapse this line onto the previous one

zerothi replied:
`<80` char limit

mattip commented on numpy/distutils/system_info.py:
```diff
@@ -92,20 +92,20 @@
 
 [fftw]
-fftw_libs = rfftw, fftw
-fftw_opt_libs = rfftw_threaded, fftw_threaded
```
Why did you remove this line as part of this PR?

zerothi replied:
`fftw_opt_libs` is never referenced anywhere in `system_info.py`, so I removed it due to its connection with `libraries` option

mattip replied:
interesting. Even when I go back to 575d373479 which added these lines, I do not see how it was ever parsed.

zerothi replied:
Agreed, I couldn't find any comments on it...

mattip commented on numpy/distutils/system_info.py:
```diff
@@ -92,20 +92,20 @@
 [fftw]
-fftw_libs = rfftw, fftw
-fftw_opt_libs = rfftw_threaded, fftw_threaded
-# if the above aren't found, look for {s,d}fftw_libs and {s,d}fftw_opt_libs
```
Why did you remove this comment as part of this PR?

zerothi replied:
same as for previous comment

zerothi committed:
DOC: fixed documentation in site.cfg.example

Signed-off-by: Nick Papior <nickpapior@gmail.com>

mattip commented:
LGTM. Will give other reviewers a chance to take a look before merging

mattip commented:
Thanks @zerothi
